File: serveur/test_bibliotheque.js
const assert = require("assert");
const fs = require("fs");
const os = require("os");
const path = require("path");
const B = require("./plan").bibliotheque; // LA PORTE serveur du plan

function ecrire(fichier, valeur) {
  fs.mkdirSync(path.dirname(fichier), { recursive: true });
  fs.writeFileSync(fichier, JSON.stringify(valeur), "utf-8");
}

const racine = fs.mkdtempSync(path.join(os.tmpdir(), "conseil-bibliotheque-"));
try {
  ecrire(path.join(racine, "etat", "books.json"), [{ id: "ancien" }]);
  assert.deepStrictEqual(B.charger(racine), [{ id: "ancien" }]);

  ecrire(path.join(racine, "etat", "books", "_ordre.json"), ["b", "a"]);
  ecrire(path.join(racine, "etat", "books", "a.json"), { id: "a" });
  ecrire(path.join(racine, "etat", "books", "b.json"), { id: "b" });
  assert.deepStrictEqual(B.charger(racine).map((x) => x.id), ["b", "a"]);
  assert.strictEqual(B.cheminsSource(racine).length, 3);

  const une = B.ouvrir(racine);
  const deux = B.ouvrir(racine);
  une.livres.find((x) => x.id === "a").n = 1;
  deux.livres.find((x) => x.id === "b").n = 2;
  une.sauver();
  deux.sauver();
  assert.deepStrictEqual(B.charger(racine).map((x) => x.n), [2, 1]);

  ecrire(path.join(racine, "etat", "maisons.json"), [{ id: "maison-a" }]);
  const maison = path.join(racine, "etat", "maisons", "maison-a", "documents", "books");
  ecrire(path.join(maison, "_ordre.json"), ["moyens-a"]);
  ecrire(path.join(maison, "moyens-a.json"),
    { id: "moyens-a", maison_id: "maison-a", n: 0 });
  assert.deepStrictEqual(B.charger(racine).map((x) => x.id), ["b", "a", "moyens-a"]);
  const maisonSession = B.ouvrir(racine);
  maisonSession.livres.find((x) => x.id === "moyens-a").n = 7;
  maisonSession.sauver();
  assert.strictEqual(JSON.parse(fs.readFileSync(path.join(maison, "moyens-a.json"))).n, 7);
  const nouveau = B.ouvrir(racine);
  nouveau.livres.push({ id: "nouveau", maison_id: "maison-a" });
  nouveau.sauver();
  assert.deepStrictEqual(JSON.parse(fs.readFileSync(path.join(maison, "_ordre.json"))),
    ["moyens-a", "nouveau"]);

  const trois = B.ouvrir(racine);
  const quatre = B.ouvrir(racine);
  trois.livres.find((x) => x.id === "a").n = 3;
  quatre.livres.find((x) => x.id === "a").n = 4;
  trois.sauver();
  assert.throws(() => quatre.sauver(), /volume modifié/);
  assert.deepStrictEqual(
    fs.readdirSync(path.join(racine, "etat", "books"))
      .filter((n) => n.endsWith(".tmp") || n.startsWith(".bibliotheque-")),
    []);

  fs.unlinkSync(path.join(racine, "etat", "books", "b.json"));
  assert.throws(() => B.charger(racine), /volume absent/);
  console.log("bibliotheque.js: OK");
} finally {
  fs.rmSync(racine, { recursive: true, force: true });
}

File: scripts/test_bibliotheque.py
# -*- coding: utf-8 -*-
import json
import os
import tempfile
import unittest

import scripts.bibliotheque as B


def ecrire_json(chemin, valeur):
    os.makedirs(os.path.dirname(chemin), exist_ok=True)
    with open(chemin, "w", encoding="utf-8") as f:
        json.dump(valeur, f, ensure_ascii=False)


class BibliothequeTest(unittest.TestCase):
    def test_le_monolithe_reste_le_repli(self):
        with tempfile.TemporaryDirectory() as etat:
            ecrire_json(os.path.join(etat, "books.json"), [{"id": "a"}])
            self.assertEqual([{"id": "a"}], B.charger(etat))
            self.assertFalse(B.est_scindee(etat))

    def test_le_manifeste_active_les_volumes_et_leur_ordre(self):
        with tempfile.TemporaryDirectory() as etat:
            ecrire_json(os.path.join(etat, "books.json"), [{"id": "ancien"}])
            ecrire_json(os.path.join(etat, "books", "_ordre.json"), ["b", "a"])
            ecrire_json(os.path.join(etat, "books", "a.json"), {"id": "a"})
            ecrire_json(os.path.join(etat, "books", "b.json"), {"id": "b"})
            self.assertEqual(["b", "a"], [x["id"] for x in B.charger(etat)])
            self.assertTrue(B.est_scindee(etat))

    def test_un_volume_absent_ne_retombe_pas_sur_le_monolithe(self):
        with tempfile.TemporaryDirectory() as etat:
            ecrire_json(os.path.join(etat, "books.json"), [{"id": "ancien"}])
            ecrire_json(os.path.join(etat, "books", "_ordre.json"), ["absent"])
            with self.assertRaisesRegex(B.BibliothequeInvalide, "volume absent"):
                B.charger(etat)

    def test_le_fichier_doit_dire_son_propre_id(self):
        with tempfile.TemporaryDirectory() as etat:
            ecrire_json(os.path.join(etat, "books", "_ordre.json"), ["a"])
            ecrire_json(os.path.join(etat, "books", "a.json"), {"id": "b"})
            with self.assertRaisesRegex(B.BibliothequeInvalide, "porte l'id"):
                B.charger(etat)

    def test_deux_sessions_peuvent_ecrire_deux_volumes_distincts(self):
        with tempfile.TemporaryDirectory() as etat:
            ecrire_json(os.path.join(etat, "books", "_ordre.json"), ["a", "b"])
            ecrire_json(os.path.join(etat, "books", "a.json"), {"id": "a", "n": 0})
            ecrire_json(os.path.join(etat, "books", "b.json"), {"id": "b", "n": 0})
            une, deux = B.ouvrir(etat), B.ouvrir(etat)
            une.livres[0]["n"] = 1
            deux.livres[1]["n"] = 2
            une.sauver()
            deux.sauver()
            self.assertEqual([1, 2], [x["n"] for x in B.charger(etat)])

    def test_deux_sessions_ne_recouvrent_pas_le_meme_volume(self):
        with tempfile.TemporaryDirectory() as etat:
            ecrire_json(os.path.join(etat, "books", "_ordre.json"), ["a"])
            ecrire_json(os.path.join(etat, "books", "a.json"), {"id": "a", "n": 0})
            une, deux = B.ouvrir(etat), B.ouvrir(etat)
            une.livres[0]["n"] = 1
            deux.livres[0]["n"] = 2
            une.sauver()
            with self.assertRaisesRegex(B.BibliothequeModifiee, "volume modifié"):
                deux.sauver()
            self.assertEqual(1, B.charger(etat)[0]["n"])

    def test_le_monolithe_refuse_toute_ecriture_concurrente(self):
        with tempfile.TemporaryDirectory() as etat:
            ecrire_json(os.path.join(etat, "books.json"), [{"id": "a", "n": 0}])
            une, deux = B.ouvrir(etat), B.ouvrir(etat)
            une.livres[0]["n"] = 1
            deux.livres[0]["n"] = 2
            une.sauver()
            with self.assertRaisesRegex(B.BibliothequeModifiee, "a changé"):
                deux.sauver()


if __name__ == "__main__":
    unittest.main()
